src/main/kotlin/org/rust/lang/core/types/infer/MemoryCategorization.kt

```diff
@@ -5,57 +5,167 @@
+    data class Downcast(val cmt: Cmt, val element: RsElement) : Categorization()
+}
+
+sealed class BorrowKind {
+    object ImmutableBorrow : BorrowKind()
+    object MutableBorrow : BorrowKind()
+
+    companion object {
+        fun from(mutability: Mutability): BorrowKind =
+            when (mutability) {
+                Mutability.IMMUTABLE -> ImmutableBorrow
+                Mutability.MUTABLE -> MutableBorrow
+            }
+    }
+}
+
+sealed class PointerKind {
+    data class BorrowedPointer(val borrowKind: BorrowKind, val region: Region) : PointerKind()
+    data class UnsafePointer(val mutability: Mutability) : PointerKind()
+}
+
+sealed class InteriorKind {
+    object InteriorField : InteriorKind()
+    object InteriorElement : InteriorKind()
+}
+
+sealed class ImmutabilityBlame {
+    class ImmutableLocal(val element: RsElement) : ImmutabilityBlame()
+    object ClosureEnv : ImmutabilityBlame()
+    class LocalDeref(val element: RsElement) : ImmutabilityBlame()
+    object AdtFieldDeref : ImmutabilityBlame()
+}
+
+/** Mutability of the expression address */
 enum class MutabilityCategory {
+    /** Any immutable */
     Immutable,
     /** Directly declared as mutable */
     Declared,
     /** Inherited from the fact that owner is mutable */
     Inherited;
 
     companion object {
-        fun valueOf(mutability: Mutability): MutabilityCategory =
+        fun from(mutability: Mutability): MutabilityCategory =
             when (mutability) {
-                Mutability.IMMUTABLE -> MutabilityCategory.Immutable
-                Mutability.MUTABLE -> MutabilityCategory.Declared
+                Mutability.IMMUTABLE -> Immutable
+                Mutability.MUTABLE -> Declared
+            }
+
+        fun from(borrowKind: BorrowKind): MutabilityCategory =
+            when (borrowKind) {
+                is ImmutableBorrow -> Immutable
+                is MutableBorrow -> Declared
+            }
+
+        fun from(pointerKind: PointerKind): MutabilityCategory =
+            when (pointerKind) {
+                is BorrowedPointer -> from(pointerKind.borrowKind)
+                is UnsafePointer -> from(pointerKind.mutability)
             }
     }
 
     fun inherit(): MutabilityCategory =
         when (this) {
-            MutabilityCategory.Immutable -> MutabilityCategory.Immutable
-            MutabilityCategory.Declared, MutabilityCategory.Inherited -> MutabilityCategory.Inherited
+            Immutable -> Immutable
+            Declared, Inherited -> Inherited
         }
 
     val isMutable: Boolean
         get() = when (this) {
-            MutabilityCategory.Immutable -> false
-            MutabilityCategory.Declared, MutabilityCategory.Inherited -> true
+            Immutable -> false
+            Declared, Inherited -> true
         }
 }
 
 /**
- * Category, MutabilityCategory, and Type
- * Currently supports only MutabilityCategory and Type
+ * [Cmt]: Category, MutabilityCategory, and Type
```

Comment:
🤔

Reply:
Are you concerned about the name `Cmt`? Do you have better name for it?

Reply:
`ExprInfo`? 🤔

```Kotlin
val RsExpr.info: ExprInfo? get() = processExpr(this)
```
looks better than
```Kotlin
val RsExpr.cmt: Cmt? get() = processExpr(this)
```
IMO

Reply:
`ExprInfo` is a quite ambiguous name. Also `cmt` (as parameter/variable) is used widely in rustc (as well as `ty`).